src/org/pentaho/reporting/platform/plugin/async/IAsyncReportListener.java

```diff
@@ -2,7 +2,7 @@
 
 import org.pentaho.reporting.engine.classic.core.event.ReportProgressListener;
 
-public interface IAsyncReportListener extends ReportProgressListener {
+public interface IAsyncReportListener extends ReportProgressListener, IAsyncReportState {
 
```

Comment:
More explanation: Your listener is created in the "PentahoAsyncExecutor" and only references properties from the task. So logically, there is no need to let the "PentahoAsyncExecutor" create it. If you move the creation of the listener into your "PentahoAsyncReportExecution", then you can eliminate the setListener(..) method and you can be sure about how your *implementation* computes the state (and how to get it). 

The only thing you need is a UniqueID for the key. Funny enough, your PentahoAsyncReportExecution class has such an ID, which it can share with your listener. Give your IAsyncReportExecution a getter for it, and you can use it in the key for the Executor.